File: apps/api/src/modules/kyc/rules/kyc-handoff.ts
/**
 * R85 — Passage de main entre validateurs, section par section.
 *
 * Le validateur passe la main à l'étape suivante (nextStep) ou rend la main à
 * l'étape précédente (revenir). MESSAGE OBLIGATOIRE à chaque passage. Jusqu'à la
 * section de validation : valider ou rejeter (motif obligatoire, R7). Tout est
 * tracé (R48/R49) ; rien ne change d'état sans événement tracé (invariant moteur).
 * Port fidèle de olive_cpsi/kyc_handoff.py. Entité de domaine instanciée par KYC.
 */
export class HandoffError extends Error {}

export type HandoffStatus = 'en_cours' | 'valide' | 'rejete';

export interface HandoffLogEntry {
  at: string; action: string; phase: string; user: string; message: string;
  [k: string]: unknown;
}

export class KycHandoff {
  private idx = 0;
  status: HandoffStatus = 'en_cours';
  readonly log: HandoffLogEntry[] = [];
  private readonly phases: string[];

  constructor(phases: string[], idx = 0, status: HandoffStatus = 'en_cours') {
    if (!phases || phases.length === 0) throw new HandoffError('Au moins une phase requise');
    this.phases = [...phases];
    this.idx = Math.max(0, Math.min(idx, phases.length - 1));
    this.status = status;
  }

  private trace(at: Date, action: string, user: string, message: string, kw: Record<string, unknown> = {}): void {
    this.log.push({ at: at.toISOString(), action, phase: this.phases[this.idx], user, message, ...kw });
  }

  phaseCourante(): string { return this.phases[this.idx]; }
  estDerniere(): boolean { return this.idx >= this.phases.length - 1; }

  private exigeMessage(message: string, quoi: string): void {
    if (!message || !message.trim()) throw new HandoffError(`Message obligatoire ${quoi} (R85)`);
  }

  nextStep(user: string, message: string, at: Date): string {
    if (this.status !== 'en_cours') throw new HandoffError(`Dossier terminé (${this.status})`);
    this.exigeMessage(message, 'au passage de main');
    if (this.estDerniere()) throw new HandoffError('Dernière étape — utilisez valider() ou rejeter()');
    const de = this.phases[this.idx];
    this.idx += 1;
    this.trace(at, 'next_step', user, message.trim(), { de, a: this.phases[this.idx] });
    return this.phases[this.idx];
  }

  revenir(user: string, message: string, at: Date): string {
    if (this.status !== 'en_cours') throw new HandoffError(`Dossier terminé (${this.status})`);
    this.exigeMessage(message, 'au renvoi');
    if (this.idx === 0) throw new HandoffError('Déjà à la première étape');
    const de = this.phases[this.idx];
    this.idx -= 1;
    this.trace(at, 'revenir', user, message.trim(), { de, a: this.phases[this.idx] });
    return this.phases[this.idx];
  }

  valider(user: string, message: string, at: Date): HandoffStatus {
    if (this.status !== 'en_cours') throw new HandoffError(`Dossier terminé (${this.status})`);
    if (!this.estDerniere()) throw new HandoffError("La validation n'est possible qu'à la section de validation");
    this.exigeMessage(message, 'à la validation');
    this.status = 'valide';
    this.trace(at, 'validation', user, message.trim());
    return this.status;
  }

  rejeter(user: string, message: string, at: Date): HandoffStatus {
    if (this.status !== 'en_cours') throw new HandoffError(`Dossier terminé (${this.status})`);
    this.exigeMessage(message, 'au rejet'); // motif obligatoire (R7)
    this.status = 'rejete';
    this.trace(at, 'rejet', user, message.trim());
    return this.status;
  }
}

File: services/cpsi-server-py/olive_cpsi/kyc_handoff.py
# -*- coding: utf-8 -*-
"""R85 — Passage de main entre validateurs, section par section.

Le validateur d'une section passe la main à l'étape suivante (« next_step ») ou
rend la main à l'étape précédente (« revenir »). MESSAGE OBLIGATOIRE à chaque
passage — il est adressé à l'intervenant concerné (prédécesseur/successeur) et
consigné. Jusqu'à la section de validation : valider ou rejeter. Tout est tracé
(R48/R49). Rien ne change d'état sans événement tracé (invariant moteur)."""


class HandoffError(Exception):
    pass


class KycHandoff:
    def __init__(self, phases):
        if not phases:
            raise HandoffError("Au moins une phase requise")
        self.phases = list(phases)   # ex: ["ARM","RM","BRM","Compliance","Validation"]
        self.idx = 0
        self.status = "en_cours"     # en_cours | valide | rejete
        self.log = []

    def _t(self, at, action, user, message, **kw):
        self.log.append(dict(at=at.isoformat(), action=action,
                             phase=self.phases[self.idx], user=user, message=message, **kw))

    def phase_courante(self):
        return self.phases[self.idx]

    def est_derniere(self):
        return self.idx >= len(self.phases) - 1

    def _exige_message(self, message, quoi):
        if not message or not message.strip():
            raise HandoffError("Message obligatoire %s (R85)" % quoi)

    def next_step(self, user, message, at):
        if self.status != "en_cours":
            raise HandoffError("Dossier terminé (%s)" % self.status)
        self._exige_message(message, "au passage de main")
        if self.est_derniere():
            raise HandoffError("Dernière étape — utilisez valider() ou rejeter()")
        de = self.phases[self.idx]
        self.idx += 1
        self._t(at, "next_step", user, message.strip(), de=de, a=self.phases[self.idx])
        return self.phases[self.idx]

    def revenir(self, user, message, at):
        if self.status != "en_cours":
            raise HandoffError("Dossier terminé (%s)" % self.status)
        self._exige_message(message, "au renvoi")
        if self.idx == 0:
            raise HandoffError("Déjà à la première étape")
        de = self.phases[self.idx]
        self.idx -= 1
        self._t(at, "revenir", user, message.strip(), de=de, a=self.phases[self.idx])
        return self.phases[self.idx]

    def valider(self, user, message, at):
        if self.status != "en_cours":
            raise HandoffError("Dossier terminé (%s)" % self.status)
        if not self.est_derniere():
            raise HandoffError("La validation n'est possible qu'à la section de validation")
        self._exige_message(message, "à la validation")
        self.status = "valide"
        self._t(at, "validation", user, message.strip())
        return self.status

    def rejeter(self, user, message, at):
        if self.status != "en_cours":
            raise HandoffError("Dossier terminé (%s)" % self.status)
        self._exige_message(message, "au rejet")   # motif obligatoire (R7)
        self.status = "rejete"
        self._t(at, "rejet", user, message.strip())
        return self.status
